### ISSUE TITLE
Easy way to calculate replication progress

### ISSUE THREAD

**Issue by dharders** [pouchdb/pouchdb#5713]
### Issue

Replication progress feedback is required for users of large dbs. The best way I have found so far is to use `last_seq / update_seq`, as described [here](http://stackoverflow.com/questions/29138091/pouchdb-couchdb-sequence-counter-on-replicate/29143448#29143448) by @nolanlawson .

However, recently I was bitten by this technique as `update_seq` is unique to each db and replicating to an empty db does not guarantee `update_seq` will match after replication has completed. This results in a progress value that never reaches 100%, and can break other logic that is waiting for that condition.

Try as I might, I cannot get an **ACCURATE** _Progress Value_ based on the data given by the  `.on('change', ...)` event.

The values don't seem to add up. I've tried countless combinations of these stats `docs_read, docs_written, last_seq, docs.length, remote.info(), update_seq`, but I can never arrive at a 100% complete value (with certainty for all cases).

Is there a better algorithm for calculating replication progress other than `last_seq / update_seq` ? 

I'd prefer not to have to poll the /_active_tasks endpoint etc for progress too.

If a more robust algorithm is found, I would be happy to write a plugin.
### Info
- Server: CouchDB 2.0   (with `pending` docs avail in _changes feed)

**Comment by willholley** [comment]
The first problem is how you actually define complete. Unlike CouchDB, PouchDB's replicator doesn't single _changes response during replication. Instead, it reads a batch of changes and then makes another call to the remote `_changes` to get the next batch. That means there exists a race condition where the "completeness" of a replication might change over the course of the replication, e.g. if new changes occur in the source database.

As you've discovered sequence numbers aren't a reliable measure, particularly in Couch 2.0 where the integer component of the sequence is somewhat arbitrary. This is why the `pending` field in `_changes` exists, and this should produce the most accurate number of remaining changes to process. Note that this number doesn't necessarily match the number of rows returned in the feed (e.g. if there is a filter specified, pending might be 100 but, if only 1 doc passes the filter, Pouch would only receive one row in the response before it completes).

My suggestion would be to rely only on the `pending` field in changes (which we'd need to surface in PouchDB) and track this value between batches. As mentioned above, this might be difficult to track as a % complete because the total changes might change over time, but as an approximation, you could record the first `pending` value returned during a replication call and assume that is the total.

**Comment by dharders** [comment]
Thanks for your great answer @willholley 

I was thinking the same thing with capturing the first `pending` value and adding that to the `batch_size` to equal the 'total_changes'. Then the 'progress' would be calculated from `pending / total_changes`.

So I guess the TLDR of this is wait until `pending` is available in pouchdb and the .on('change') event ?

**Comment by dharders** [comment]
Thinking this should handle most scenarios (filtered replication needs some more thought).

// assumes `pending` is available in the .on('change', (info) => { // info.pending}) event

``` javascript
var pendingMax = 0;
var batch_size = 1000;    // must match your replication options

function getProgress(pending) {
  var progress;
  pendingMax = pendingMax < pending ? pending + batch_size : pendingMax;  // first time capture
  if (pendingMax > 0) {
    progress = 1 - pending/pendingMax;
    if (pending === 0) {
      pendingMax = 0;    // reset for live/next replication
    }
  } else {
    progress = 1;  // 100%
  }
  return progress;
}
```

Example use:

``` javascript
var replication = db.replicate.from(remote, {
  live: true,
  retry: true,
  batches_limit: 10,
  batch_size: 1000      // must match above batch_size
}).on('change', function (info) {
  console.log('Replication Progress', getProgress(info.pending));
})
...
```

**Comment by willholley** [comment]
I think that's pretty much right. I wouldn't use batch size in the calculation - the only thing you can reliably measure is the pending delta. Maybe just change:

```
pendingMax = pendingMax < pending ? pending + batch_size : pendingMax;  // first time capture
```

to 

```
pendingMax = pendingMax < pending ? pending : pendingMax;  // first time capture
```

**Comment by dharders** [comment]
I only included `batch_size` to gain an extra step of precision and it also avoided the first _change_ batch to equal 0% (where obviously we had progressed > 0%). 

I assumed `batch_size` was safe to use here because CouchDB seemed to honour this value (unless you're alluding to filtered rep etc, which I haven't used yet) and if we have less than `batch_size` docs to replicate, `pending` would be 0, so we return 100% anyway. 

But I guess it's not a huge issue without `batch_size` if we gain reliability.

**Comment by jareware** [comment]
> This results in a progress value that never reaches 100%, and can break other logic that is waiting for that condition.

I think a more reliable way to detect the *completion* of the replication is to use the Promise returned by `replicate()`.

**Comment by dharders** [comment]
@jareware yes that's true for one-shot replication but you don't get that benefit for **live** replication, which I primarily use. Sorry, should have made that more clear.

**Comment by jareware** [comment]
What I've actually done is first run a one-off replication and once that finishes, start the continuous one. There might be a bit of overhead, but on the other hand makes it super clear in the code which is which, and what change-events mean.

**Comment by dharders** [comment]
@jareware 

> first run a one-off replication and once that finishes, start the continuous one

The issue we're trying to solve is the fact that `last_seq / update_seq` is not guaranteed to equal 100%.

> The update seq number is specific to the database. Suppose you have 9 documents in a database and the update_seq for it is 17. update_seq indicates that the database has been updated 17 times even though the number of documents are 9. Now if you replicate it to a new database which is empty then the update_seq of the new database would be 9 because it has only been updated 9 times. Since the new database was empty only 9 new documents were created.
> 
> If it contained a few documents then the update number would have been different depending upon what their _rev's were. So update_seq can't be relied upon to calculate progress of replication operation.

So even after doing a 'one-off' replication then switch to 'continuous' (which in fact, is what I already do) the progress calculation __sometimes__ never reaches 100% using the current 'best' method. We also can't use the `completed` event to signal 100% progress with continuous replication as it's only ever fired when replication is 'cancelled'.

This should all be fixed when `pending` lands in PouchDB (see #5710).

**Comment by jareware** [comment]
Can't you just force the progress to 100% once the one-off replication finishes? Sorry if I'm missing something obvious.

**Comment by dharders** [comment]
Picture this:

A user (who has been offline for a while and has 1000's of changes) connects and replication begins. Now consider another user connected to the same remote db is using **live** replication. The progress is calculated per change **from given values** (ie last_seq/update_seq). I've personally experienced issues where last_seq is less than update_seq and the gap grows with Db size. So relying on this technique is error prone. I've had replication finish at 95% and less with larger dbs. 

Also, how do we define finished in **live** replication? Both replicateTo and replicateFrom 'paused' without error? Paused for how long? Considered slow mobile connections?

There are many 'hacks' we could apply, but this issue is about ACCURATE and consistent progress calculation. When #5710 lands, we should have it. 

If you want to discuss it further, maybe PM me as this thread is already getting too long ;)

**Comment by daleharvey** [comment]
For live replication, `paused` should be fired without error at the instant that the replication is complete, if you see that even it means the replication is complete and we are now awaiting future changes (in the case of sync it fires when both replications are complete)

**Comment by dharders** [comment]
@daleharvey Hmmm, in the past I've experienced `paused` events firing then straight after back to `active` then `paused` -> `active` etc. (in a somewhat thrashing) hence the timeout to check if we're still paused without change. I just assumed this was normal replication behaviour. Maybe this issue has been fixed in a recent version, but last time I tested it was still present (I'll have to check the version I was using)

**Comment by daleharvey** [comment]
Yeh we have had issues with `pause` thrashing before, theres definitely fixes in there and hopefully its all fixed, its a bug for pouch to fire paused (without error) unless it knows its complete

**Comment by dharders** [comment]
@daleharvey ok great. Thanks for the confirmation of correct replication event behaviour. I'll update pouchdb to latest version and do some more testing. 

If the thrashing behaviour is gone then I guess this issue could be considered closed with `last_seq / update_seq` best current solution until #5710 (pending property) lands.

**Comment by nolanlawson** [comment]
Sounds like we need a generic and simple way to calculate replication progress. This could potentially be implemented as a plugin, but someone ought to implement it.

**Comment by sagarkm** [comment]
I tried with the latest build of PouchDb 6.1.2 with {live: true, retry: true} and I have observed on calling sync, we get "pause" event first (without any error), then it goes into Active and then "pause" again without any error. 
Curent event flow: Pause -> Active -> Pause
Ideally, it should be Active -> Pause

Is there a plugin to show the sync status or pending property added so far?

**Comment by Mohsen7s** [comment]
This would a very useful feature. No need for extra plugin  @nolanlawson Why not add to pouch-db built in featues ?

**Comment by jacwright** [comment]
Can we get the `pending` property exposed in pouchdb? Would a plugin be able to get this, or would it need to be handled in the core?

**Comment by willholley** [comment]
@jacwright I don't think it would be difficult to do for the http adapter only - possibly just modifying the file you suggest, adding appropriate docs and a basic test. Do you want to have a stab at a PR for that?

**Comment by dharders** [comment]
@jacwright @willholley  I just started work on supporting the `pending` property, see PR #6812 

Would love some extra eyes on it (and if anyone can confirm if travis errors are due to general travis flakeyness or genuine bug(s)) 

Also, I created another issue/feature request #6818 if you wouldn't mind checking out. It essentially is a PouchDB created `pending` value of docs to UPLOAD. It's required to have full replication progress reporting in both directions. The `pending` property given by CouchDb 2.0+ is only pending docs to DOWNLOAD from remote. I would love some suggestions on the best way to go about emulating it on the frontend and if either of you are familiar with how CouchDb does it on the server (I haven't dug into their source yet!).

Cheers.
